core/scale/encoder/primitives.hpp

```diff
@@ -0,0 +1,434 @@
+    const size_t size = value.backend().size();
+    const mp::limb_type *const p = value.backend().limbs();
+
+    constexpr size_t limb_sz = sizeof(mp::limb_type);
+    size_t ix = 0ull;
+    for (; ix < size; ++ix) {
+      *(mp::limb_type *)&result[1ull + ix * limb_sz] = math::toLE(p[ix]);
+    }
+
+    putByte(func, result, bigIntLength + 1ull);
+  }
+
+  template <
+      typename F,
+      typename It,
+      typename = std::enable_if_t<
+          !std::is_same_v<typename std::iterator_traits<It>::value_type, void>>>
+  constexpr void encode(const F &func, It begin, It end) {
+    while (begin != end) {
+      encode(func, *begin);
+      ++begin;
+    }
+  }
+
+  template <typename F, typename T, ssize_t S>
+  constexpr void encode(const F &func, const gsl::span<T, S> &c) {
+    if constexpr (S == -1) {
+      encodeCompact(func, c.size());
+      encode(func, c.begin(), c.end());
+    } else {
+      using E = std::decay_t<T>;
+      if constexpr (std::is_integral_v<E> && sizeof(E) == 1u) {
+        putByte(func, c.data(), c.size());
+      } else {
+        for (const auto &e : c) {
+          encode(func, e);
+        }
+      }
+    }
+  }
+
+  template <typename F, typename T, size_t size>
+  constexpr void encode(const F &func, const std::array<T, size> &c) {
+    for (const auto &e : c) {
+      encode(func, e);
+    }
+  }
+
+  template <typename F, typename K, typename V>
+  constexpr void encode(const F &func, const std::map<K, V> &c) {
+    encodeCompact(func, c.size());
+    encode(func, c.begin(), c.end());
+  }
+
+  template <typename FN, typename F, typename S>
+  constexpr void encode(const FN &func, const std::pair<F, S> &p) {
+    encode(func, p.first);
+    encode(func, p.second);
+  }
+
+  template <typename F, typename T>
+  constexpr void encode(const F &func, const std::vector<T> &c) {
+    encodeCompact(func, c.size());
+    encode(func, c.begin(), c.end());
+  }
+
+  template <typename F, typename T>
+  constexpr void encode(const F &func, const std::shared_ptr<T> &v) {
+    if (v == nullptr) {
+      raise(::scale::EncodeError::DEREF_NULLPOINTER);
+    }
+    encode(func, *v);
+  }
+
+  template <typename F, typename T>
+  constexpr void encode(const F &func, const std::unique_ptr<T> &v) {
+    if (v == nullptr) {
+      raise(::scale::EncodeError::DEREF_NULLPOINTER);
+    }
+    encode(func, *v);
+  }
+
+  template <typename F, typename T>
+  constexpr void encode(const F &func, const std::list<T> &c) {
+    encodeCompact(func, c.size());
+    encode(func, c.begin(), c.end());
+  }
+
+  template <typename F, typename T>
+  constexpr void encode(const F &func, const std::deque<T> &c) {
+    encodeCompact(func, c.size());
+    encode(func, c.begin(), c.end());
+  }
+
+  template <typename F, typename... Ts>
+  constexpr void encode(const F &func, const std::tuple<Ts...> &v) {
+    if constexpr (sizeof...(Ts) > 0) {
+      std::apply([&](const auto &...s) { (..., encode(func, s)); }, v);
```

Comment:
Is `std::apply` better than `encode<N=0>(variant) { encode<N+1>() }` (or maybe `encode(variant) { encode(variant.which()); visit(variant, encode); }`?